From 2a9beb96c181c9e9d941c430800e193490e0394c Mon Sep 17 00:00:00 2001
From: Ewan Milne <emilne@redhat.com>
Date: Wed, 20 Aug 2014 18:58:42 -0400
Subject: [scsi] use the scsi data buffer length to extract transfer size

Message-id: <1408561176-18133-4-git-send-email-emilne@redhat.com>
Patchwork-id: 88422
O-Subject: [RHEL7.1 PATCH e-stor 03/57] [SCSI] use the scsi data buffer length to extract transfer size
Bugzilla: 1132092
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: "Ewan D. Milne" <emilne@redhat.com>

upstream commit 5616b0a46ed82eb9a093f752fc4d7bd3cc688583

Tested by Ewan D. Milne and Barry Donahue with a combination
of manual and automated testing.  No issues seen.

upstream commit 5616b0a46ed82eb9a093f752fc4d7bd3cc688583
Author: Martin K. Petersen <martin.petersen@oracle.com>
Date:   Tue Jun 24 16:59:35 2014 +0200

    [SCSI] use the scsi data buffer length to extract transfer size

    Commit 8846bab180fa introduced a helper that can be used to query the
    wire transfer size for a SCSI command taking protection information into
    account.

    However, some commands do not have a 1:1 mapping between the block range
    they work on and the payload size (discard, write same). After the
    scatterlist has been set up these requests use __data_len to store the
    number of bytes to report completion on. This means that callers of
    scsi_transfer_length() would get the wrong byte count for these types of
    requests.

    To overcome this we make scsi_transfer_length() use the scatterlist
    length in the scsi_data_buffer as basis for the wire transfer
    calculation instead of __data_len.

    Reported-by: Christoph Hellwig <hch@infradead.org>
    Debugged-by: Mike Christie <michaelc@cs.wisc.edu>
    Signed-off-by: Martin K. Petersen <martin.petersen@oracle.com>
    Signed-off-by: Christoph Hellwig <hch@lst.de>
    Reviewed-by: Sagi Grimberg <sagig@mellanox.com>
    Fixes: d77e65350f2d82dfa0557707d505711f5a43c8fd
    Cc: stable@vger.kernel.org
    Signed-off-by: James Bottomley <JBottomley@Parallels.com>

Signed-off-by: Ewan D. Milne <emilne@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/scsi/scsi_cmnd.h b/include/scsi/scsi_cmnd.h
index c3e087b..024c4b4 100644
--- a/include/scsi/scsi_cmnd.h
+++ b/include/scsi/scsi_cmnd.h
@@ -320,7 +320,7 @@ static inline void set_driver_byte(struct scsi_cmnd *cmd, char status)
 
 static inline unsigned scsi_transfer_length(struct scsi_cmnd *scmd)
 {
- unsigned int xfer_len = blk_rq_bytes(scmd->request);
+ unsigned int xfer_len = scsi_out(scmd)->length;
  unsigned int prot_op = scsi_get_prot_op(scmd);
  unsigned int sector_size = scmd->device->sector_size;
 
-- 
1.7.1